src/mumble/PipeWire.h

```diff
@@ -0,0 +1,127 @@
+// Copyright 2021 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+#ifndef MUMBLE_MUMBLE_PIPEWIRE_H
+#define MUMBLE_MUMBLE_PIPEWIRE_H
+
+#include "AudioInput.h"
+#include "AudioOutput.h"
+
+#include <QWaitCondition>
+
+class PipeWireInit;
+
+struct pw_buffer;
+struct pw_loop;
+struct pw_properties;
+struct pw_stream;
+struct pw_stream_events;
+struct pw_thread_loop;
+
+struct spa_dict;
+struct spa_pod;
+
+class PipeWireEngine {
+private:
+	Q_DISABLE_COPY(PipeWireEngine)
+protected:
+	bool m_ok;
+	pw_loop *m_loop;
+	pw_stream *m_stream;
+	pw_thread_loop *m_thread;
+	std::unique_ptr< pw_stream_events > m_events;
+
+public:
+	inline bool isOk() { return m_ok; };
+
+	bool connect(const uint8_t direction, const uint32_t *channels, const uint8_t nChannels);
+
+	void start();
+	void stop();
+
+	pw_buffer *dequeueBuffer();
+	void queueBuffer(pw_buffer *buffer);
+
+	PipeWireEngine(const char *category, void (*callback)(void *data));
+	~PipeWireEngine();
+};
```

Comment:
Something that I always ignored until recently is that according to our coding guidelines (somewhere on the wiki) we want to have the sections in order public, protected private as a reader is usually most interested in public fields.
I think this kinds makes sense, so maybe you also want to adapt that style for new classes :)

Reply:
Excellent point. In this case I would remove the `private` label and keep `Q_DISABLE_COPY` at the top though.

Reply:
I changed idea as soon as I noticed how horrible `PipeWireSystem` would look because of the `friend` directives.